src/QsCompiler/Transformations/CallGraph/ConcreteCallGraphWalker.cs

```diff
@@ -0,0 +1,275 @@
+                    {
+                        if (typeParam is QsLocalSymbol.ValidName name)
+                        {
+                            if (request.ParamResolutions.TryGetValue(Tuple.Create(callable.FullName, name.Item), out var res))
+                            {
+                                return res;
+                            }
+                            else
+                            {
+                                throw new ArgumentException($"Couldn't resolve all type parameters for {callable.FullName}");
+                            }
+                        }
+                        else
+                        {
+                            throw new ArgumentException("Encountered invalid type parameter name during call graph construction");
+                        }
+                    }).ToImmutableArray());
+                    typeArgs = SpecializationBundleProperties.BundleId(typeArgs);
+                }
+                else if (relevantSpecs.Count() != 1)
+                {
+                    throw new ArgumentException($"Missing specialization {request.Kind} for {request.CallableName}");
+                }
+
+                if (!typeArgs.IsNull)
+                {
+                    // Finds the correct type specialization for the type arguments of the currentRequest.
+                    // The assumption is that upon resolution, these type arguments have been cast to
+                    // the type of any explicitly defined ones in the closest matching specialization.
+                    var specArgMatches = relevantSpecs.Select(spec =>
+                    {
+                        if (spec.TypeArguments.IsNull)
+                        {
+                            return (0, spec);
+                        }
+                        if (spec.TypeArguments.Item.Count() != typeArgs.Item.Count())
+                        {
+                            throw new ArgumentException($"Incorrect number of type arguments in request for {request.CallableName}");
+                        }
+                        var specTypeArgs = spec.TypeArguments.Item.Select(StripPositionInfo.Apply).ToImmutableArray();
+                        var mismatch = specTypeArgs.Where((tArg, idx) => !tArg.Resolution.IsMissingType && !tArg.Resolution.Equals(typeArgs.Item[idx])).Any();
+                        if (mismatch)
+                        {
+                            return (-1, spec);
+                        }
+                        var matches = specTypeArgs.Where((tArg, idx) => !tArg.Resolution.IsMissingType && tArg.Resolution.Equals(typeArgs.Item[idx])).Count();
+                        return (matches, spec);
+                    });
+
+                    if (!specArgMatches.Any(m => m.Item1 >= 0))
+                    {
+                        throw new ArgumentException($"Could not find a suitable {request.Kind} specialization for {request.CallableName}");
```

Comment:
With this program:

```qsharp
function Foo<'a>(x : 'a) : 'a {
    return x;
}

function Bar<'a>() : ('a -> 'a) {
    return Foo<'a>;
}

@EntryPoint()
function Main() : Unit {
    (Bar<Unit>())();
}
```

I get:

> System.ArgumentException: Could not find a suitable QsControlledAdjoint specialization for Graph.Foo

Reply:
This is actually the same error as above. It just gets thrown in a different place because the reference is generic. Let's consolidate this conversation on the other thread.